### DISCUSSION TITLE
Pylance removes unused import although using "# pyright: " to ignore (conftest.py)

### DISCUSSION THREAD

**Discussion by lfvjimisola** [microsoft/pylance-release#4083] [General] [answered]
I'm importing pytest fixturs in conftest.py as per [here](https://docs.pytest.org/en/latest/reference/fixtures.html#conftest-py-sharing-fixtures-across-multiple-files:). For relevance we are using pytest, black and flake8 with VS Code.

However, the import is removed on save although I seem to be following the instructions as per https://github.com/microsoft/pylance-release/issues/196.

```
from .x.y.z import (fixture1, fixture2)  # pyright: reportUnusedImport=false # type: ignore # noqa F401 
```

also tried with (as per https://github.com/microsoft/pylance-release/discussions/2691#discussioncomment-3023685):

```
from .x.y.z import (fixture1, fixture2)  # pyright: ignore[reportUnusedImport] # type: ignore # noqa F401 
```

What am I doing wrong?

**Comment by heejaechang** [comment, marked as answer]
what you need is `# pyright: ignore` on the line. the one we use for `remove unused imports` doesn't have rule name such as `reportUnusedImport`, `reportUnusedImport` is used to control reporting an error not the `remove unused import` feature. that is why your comment didn't work. and since it is not actually type related error, `# type: ignore` doesn't work.

**Reply by lfvjimisola** [reply]
@heejaechang  Thank you. My example was not complete as it was not multi-line. I think I tried that before but it didn't work with multi-line. Realized now that I need to have the `# pyright: ignore` on _EACH_ line which is different from how flake8 uses `# noqa Xnnn`.

So, the following works as expected.

```
from .reqstool.models.test_implementations import (  # noqa F401
    implementations_data,  # pyright: ignore
    git_impl_data,  # pyright: ignore
    local_impl_data,  # pyright: ignore
    maven_impl_data,  # pyright: ignore
)
```

I though that it would be enough like this (first line):

```
from .reqstool.models.test_implementations import (  # pyright: ignore # noqa F401
    implementations_data,
    git_impl_data,
    local_impl_data,
    maven_impl_data,
)
```

or this (last line)


```
from .reqstool.models.test_implementations import ( 
    implementations_data,
    git_impl_data,
    local_impl_data,
    maven_impl_data,
) # pyright: ignore # noqa F401
```

**Reply by heejaechang** [reply]
ah. yes, currently we don't have a way to disable it for whole file or whole import statement. 

can you write down behavior you want? that will help people to understand request better.

**Reply by erictraut** [reply]
If you want to ignore this diagnostic for an entire file, you can add the comment `# pyright: reportUnusedImport=false` on a line by itself at the top of the file.

Comments of the form `# type: ignore` and `# pyright: ignore` suppress diagnostics only on a single line. Refer to [this documentation](https://microsoft.github.io/pyright/#/comments) for details.

**Reply by lfvjimisola** [reply]
> ah. yes, currently we don't have a way to disable it for whole file or whole import statement.
> 
> can you write down behavior you want? that will help people to understand request better.

Ideally, both would be useful. But, having it for at least a full import as supported with "# noqa F401" would be a big improvement.

That is,

```
from .reqstool.models.test_implementations import (  # pyright: ignore # noqa F401
    implementations_data,  
    git_impl_data,  
    local_impl_data, 
    maven_impl_data,
)
```

**Reply by lfvjimisola** [reply]
> If you want to ignore this diagnostic for an entire file, you can add the comment `# pyright: reportUnusedImport=false` on a line by itself at the top of the file.
> 
> Comments of the form `# type: ignore` and `# pyright: ignore` suppress diagnostics only on a single line. Refer to [this documentation](https://microsoft.github.io/pyright/#/comments) for details.

If I understood @heejaechang correct then in my case the `# pyright: reportUnusedImport=false` for the entire file will not work either since it's used to control reporting an error not stop from the remove unused imports.

**Reply by erictraut** [reply]
Which diagnostic are you trying to suppress? When you view the diagnostic in the Problems tab, you will see the diagnostic rule associated with it. For example, with this diagnostic, you can see that it is associated with the `reportMissingImports` diagnostic rule.

<img width="549" alt="image" src="https://user-images.githubusercontent.com/7040122/227218047-8e152249-54ab-4a37-8486-7fe080417cc7.png">

**Reply by erictraut** [reply]
Incidentally, the `reportUnusedImport` rule is not enabled by default (unless you are using "strict" type checking mode), so you must have explicitly enabled this diagnostic check at some point. If you don't want to see diagnostics for this rule, you can simply disable it for your project.
